### PR TITLE
Restore pagination information

### PR DESCRIPTION
**Pull request by jonjohnsonjr**
Notably "breaking" changes here are that the registry may impose a limit on the number of tags it returns.

This was restored from b4e98332eccaf51d1974c9136caf67317bd99104, where it applied to Catalog, which was removed, but given that many registries have `adopted this behavior already, we should document it correctly.

This also restores language that clients SHOULD always use the Link header when present.

This was restored from c90b0f145ac6bc09d2636ee214486ac333edc284.

### CODE REVIEW

**Comment by sudo-bmitch** [comment]
Does this text now allow a registry to return more than `n` results? That could be an issue for clients that assume a fixed size array for the results.

**Comment by rchincha** [comment]
Also, guidelines on what the expected behavior is if the tag list changes in between, due to new images being pushed or deleted for that matter?

File: spec.md [hunk 1]
```diff
@@ -639,6 +644,57 @@ If the [referrers API](#listing-referrers) returns a 404, the client MUST fallba
...
+      ...
+    ]
+}
+```
+
+To get the _next_ `n` entries, one can create a URL where the argument `last` has the value from `tags[len(tags)-1]`.
+If there are indeed more results, the URL for the next block is encoded in an [RFC5988](https://tools.ietf.org/html/rfc5988) `Link` header, as a "next" relation.
+If the header is not present, the client can assume that all results have been received.
+
+> __NOTE:__ In the request template above, note that the brackets are required. For example, if the url is `http://example.com/v2/hello-world/tags/list?n=20&last=b`, the value of the header would be `<http://example.com/v2/hello-world/tags/list?n=20&last=b>; rel="next"`.
```
Comment (andaaron) [inline spec.md:678]: Maybe include also an example with a relative reference, such as the server returning just `/v2/hello-world/tags/list?n=20&last=b>; rel="next"`. Which I would expect to be more common, as the reply may be routed a few times until getting to the client and the ip/hostname/port the server is aware of may be different from the one the client should make the request to.

File: spec.md [hunk 2]
```diff
@@ -639,6 +644,57 @@ If the [referrers API](#listing-referrers) returns a 404, the client MUST fallba
...
+GET /v2/<name>/tags/list?n=<n from the request>&last=<last tag value from previous response>
+```
+
+The above process should then be repeated until the `Link` header is no longer set in the response.
+
+The tag list result set is represented abstractly as a lexically sorted list, where the position in that list can be specified by the query term `last`.
+The entries in the response start _after_ the term specified by `last`, up to `n` entries.
+
+The client can then issue the request with the above value from the `Link` header, receiving the values _v3_ and _v4_.
+Note that `n` may change on the second to last response or be fully omitted, depending on the server implementation.
```
[the remark this answers is not in the document]
Comment (dmesser) [reply spec.md:696]: In the second to last response, the n parameter could be adjusted by the server to match the remaining number of tags to be delivered in the last response. I would, however, say that this is more difficult for a registry to implement than just using the name of the last tag in the current result set.

**Review by dmesser — COMMENTED** [review]
Overall I like this change. For a public registry service, the wording of the current spec is essentially impossible to implement without opening an easy to exploit DoS vector:

> The response to such a request MAY return fewer than <int> results, but only when the total number of tags attached to the repository is less than <int>. Otherwise, the response MUST include <int> results.

So the proposed changes are alleviating that. The one thing I am missing here is details around how a server can/should tell the client that it requested too many tags and in which circumstances. 

In line 533 it says:

> The registry implementation MAY impose a maximum limit and return a partial set with pagination links.
See [Pagination](#pagination) for information on how to handle pagination links.

But there is no detail provided in that section about the return code. I would suggest to include the verbiage from @sudo-bmitch's earlier proposal: https://github.com/opencontainers/distribution-spec/pull/470/files#diff-bc6661da34ecae62fbe724bb93fd69b91a7f81143f2683a81163231de7e3b545R548

Allow to return a HTTP 400 response and make it so that clients rely on a reasonable minimum amount of tags to never trigger this response, like 20.

**Comment by jonjohnsonjr** [comment]
I don't think any of this is unambiguous and we should just merge it.

**Comment by dmesser** [comment]
@jonjohnsonjr Was this PR accidentally closed, or did you have something else in mind?

**Comment by tianon** [comment]
I'm interpreting between the lines a little bit (so massive apologies to Jon if I'm inaccurate here), but I think it was probably closed because this isn't actually a _change_ to the spec, but restoring content that _was_ in the spec and got removed during refactoring (so he probably didn't want to spend too many cycles discussing/nit picking the previous behavior/design choices that are already implemented in many registries, and documenting that interop is really why this spec exists :sweat_smile:).

**Comment by dmesser** [comment]
Got it.  Has this revert happened yet? Because the spec in its current form is currently really problematic with regards to pagination: https://github.com/opencontainers/distribution-spec/blob/main/spec.md?plain=1#L532-L533

> <name> is the namespace of the repository, and <int> is an integer specifying the number of tags requested. The response to such a request MAY return fewer than <int> results, but only when the total number of tags attached to the repository is less than <int>. Otherwise, the response MUST include <int> results. 

This essentially opens up a DoS vector if the client can expect to put arbitrarily high values in `<int>` and the registry is expected to respond with it, with no means of pushing back with an error.

**Comment by jdolitsky** [comment]
Can we re-address this with a fresh PR and then afterwards cut a 1.1.1?

**Pull request closed without merging**